proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+Assert(c.Get() == 7);
+```
+
+Normally an argument to a non-`ref` parameter should not be marked `ref`, but it
+is allowed in a generic context where the parameter may sometimes be `ref`.
+
+Operators will mostly not take `ref` parameters, with these exceptions:
+
+-   [the address-of operator](/docs/design/expressions/pointer_operators.md)
+    `&`;
+-   [the indexing operator](/docs/design/expressions/indexing.md) `[`...`]`;
+-   [the member access operator](/docs/design/expressions/member_access.md) `.`;
+    and
+-   [the assignment operators](/docs/design/assignment.md) such as `=`, `+=`,
+    and `++`.
+
+In the cases that they do, observe that the `ref` parameter will be the `self`
+parameter, and so will not be marked with `ref` at the call site.
+
+As an _experiment_, we are saying a pointer formed by taking the address of a
+`ref` bound name is LLVM-`captures(none)` and LLVM-`noalias`. This means that
+while a `ref` parameter could be passed into a function by address, the
+restrictions also allow a "move-in-move-out" approach (once we define the move
+operation), assuming it is not [marked `bound`](#bound-parameters). The intent
+here is to leave the door open to a calling convention using registers and less
+indirection for small-enough objects.
+
+This means that the following code is invalid:
+
+```carbon
+fn F(ref a: i32, ref b: i32) -> bool;
+
+fn G() -> bool {
+  var v: i32 = 1;
+  return F(ref v, ref v);
+}
+```
+
+### `ref`, `let`, and `var` returns
+
+The return of a function can optionally be marked `ref`, `let`, or `var`. These
+control the category of the call expression invoking the function, and how the
+return expression is returned.
+
+```carbon
+var global: i32 = 2;
+fn ReturnRef() -> ref i32 {
+  // ❌ Invalid: return 2;
+
+  // ✅ Valid: return a reference expression with
+  // sufficient lifetime.
+  return global;
+}
+// Call `ReturnRef` and use the resulting reference.
+ReturnRef() += 3;
+Assert(global == 5);
+
+// Result of `ReturnRef` can be bound using a `ref`
+// binding.
+fn AddFive() {
+  let ref r: i32 = ReturnRef();
+  r += 5;
+}
+AddFive();
+Assert(global == 10);
+
+fn ReturnLet() -> let i32 {
+  return 2;
+}
+// ReturnLet() is a value expression.
+let l: i32 = ReturnLet();
+
+fn ReturnVar() -> var i32 {
+  return 2;
+}
+// ReturnVar() is an initializing expression.
+var j: i32 = ReturnVar();
+```
+
+-   A call to a function declared `-> ref T` is a reference expression. The
+    function will return the address of a `T` object.
+-   A call to a function declared `-> let T` is a value expression. The function
+    will return the value representation of `T`. Since values have no address,
+    the value representation may be returned in registers.
+-   A call to a function declared `-> var T` is an initializing expression. The
+    caller provides the address of storage to initialize with an object of type
+    `T`, that the caller owns upon return. The object will never be returned in
+    registers.
```

Comment:
After chatting with @zygoloid, I think I convinced him that we could fold together `-> T` and `-> var T`. I'm happy to write up more fully the idea, or discuss it / take notes on it, but leaving a quick comment here so I don't forget.

For my own memory, the most convincing thing was to start from thinking about `var` *parameters*, and how we would want those to work for an owning pointer like `std::unique_ptr`.

More details when I have a bit more time at a keyboard...